calculator.go

```diff
@@ -1,13 +1,90 @@
 // Package calculator provides a library for simple calculations in Go.
 package calculator
 
+import (
+	"errors"
+	"regexp"
+	"strconv"
+)
+
+var evalRX *regexp.Regexp = regexp.MustCompile(`^\s*(\d+\.?\d*)\s*([^\d\.])\s*(\d+\.?\d*)\s*$`)
+
 // Add takes two numbers and returns the result of adding them together.
-func Add(a, b float64) float64 {
-return a + b
+func Add(a float64, b ...float64) float64 {
+	for _, bb := range b {
+		a += bb
+	}
+	return a
 }
 
 // Subtract takes two numbers and returns the result of subtracting the second
 // from the first.
-func Subtract(a, b float64) float64 {
-	return b - a
+func Subtract(a float64, b ...float64) float64 {
+	for _, bb := range b {
+		a -= bb
+	}
+	return a
+}
+
+// Multiply takes two numbers and returns the result of multiplying them
+func Multiply(a float64, b ...float64) float64 {
+	for _, bb := range b {
+		a *= bb
+	}
+	return a
+}
+
+// Divide takes two numbers and divides the first one by the second one
+func Divide(a float64, b ...float64) (float64, error) {
+	for _, bb := range b {
+		if bb == 0 {
+			return 0, errors.New("division by zero")
+		}
+		a /= bb
+	}
+	return a, nil
+}
+
+// Sqrt returns the square root of a positive number, an error otherwise.
+func Sqrt(a float64) (float64, error) {
+	if a < 0 {
+		return 0, errors.New("square root of a negative number")
+	}
+	x, y := a/2, 0.0
+	for i := 0; i < 5000; i++ {
+		if x == 0 {
+			break
+		}
+		y = 0.5 * (x + (a / x))
+		x = y
+	}
+
+	return y, nil
+}
+
+//Evaluate evaluates a given string expression and returns the result or an error
+func Evaluate(e string) (float64, error) {
+	m := evalRX.FindStringSubmatch(e)
+	if len(m) == 0 {
+		return 0, errors.New("invalid expression")
+	}
+	a, err := strconv.ParseFloat(m[1], 64)
+	if err != nil {
+		return 0, err
+	}
+	b, err := strconv.ParseFloat(m[3], 64)
+	if err != nil {
+		return 0, err
+	}
+	switch m[2] {
+	case "+":
+		return Add(a, b), nil
+	case "-":
+		return Subtract(a, b), nil
+	case "*":
+		return Multiply(a, b), nil
+	case "/":
+		return Divide(a, b)
+	}
+	return 0, errors.New("invalid operation")
```

Comment:
Matt Ryer would be furious! I'm curious if you think there's a better way to write this function? Still I have to admit that I don't hate it!

Reply:
Why do you think Mat would object to this?

Reply:
Yeah sorry that reads a bit like a youtube comment.

Well it's not part of "the happy path" but it's on the left-most line. Anyway following your advise of moving this return to the default case and using a nil error should takes care of that.

Reply:
Yes, I see what you mean. The idea of left-aligning the happy path has been a strong influence on me—I'd been programming for at least 35 years before I ever came across it, but now I've adopted it wholeheartedly. It makes functions so much easier to read.